File: iOS/dimina/DiminaKit/Container/Api/Storage/StorageAPI.swift
//
//  StorageAPI.swift
//  dimina
//
//  Created by DosLin on 2025/5/10.
//

import Foundation

/**
 * Storage API implementation
 *
 * Handles data storage operations like setting, getting, and removing stored data
 */
public class StorageAPI: DMPContainerApi {
    
    // API method names
    private static let SET_STORAGE_SYNC = "setStorageSync"
    private static let GET_STORAGE_SYNC = "getStorageSync"
    private static let REMOVE_STORAGE_SYNC = "removeStorageSync"
    private static let CLEAR_STORAGE_SYNC = "clearStorageSync"
    private static let SET_STORAGE = "setStorage"
    private static let GET_STORAGE = "getStorage"
    private static let REMOVE_STORAGE = "removeStorage"
    private static let CLEAR_STORAGE = "clearStorage"
    private static let GET_STORAGE_INFO_SYNC = "getStorageInfoSync"
    private static let GET_STORAGE_INFO = "getStorageInfo"
    
    // 在类初始化时确保DMPStorage已初始化
    override public init(app: DMPApp? = nil) {
        super.init(app: app)
        DMPStorage.shared.initialize()
    }
    
    // Set storage synchronously
    @BridgeMethod(SET_STORAGE_SYNC)
    var setStorageSync: DMPBridgeMethodHandler = { param, env, callback in
        let param = param.getMap()
        guard let key = param.get("key") as? String else { return false }
        let data = param.get("data")
        let encrypt = param.get("encrypt") as? Bool ?? false
        
        guard let data = data else { return false }
        
        let result = DMPStorage.shared.set(key: key, value: data, encrypted: encrypt)
        return result
    }
    
    // Get storage synchronously
    @BridgeMethod(GET_STORAGE_SYNC)
    var getStorageSync: DMPBridgeMethodHandler = { param, env, callback in
        guard let key = param.getValue() as? String else { return }
        let value = DMPStorage.shared.get(key: key, encrypted: false)
        return DMPBridgeParam(value: value)
    }
    
    // Remove storage synchronously
    @BridgeMethod(REMOVE_STORAGE_SYNC)
    var removeStorageSync: DMPBridgeMethodHandler = { param, env, callback in
        let param = param.getMap()
        guard let key = param.get("key") as? String else { return false }
        let encrypt = param.get("encrypt") as? Bool ?? false
        
        DMPStorage.shared.remove(key: key, encrypted: encrypt)
        return true
    }
    
    // Clear storage synchronously
    @BridgeMethod(CLEAR_STORAGE_SYNC)
    var clearStorageSync: DMPBridgeMethodHandler = { param, env, callback in
        // 清除所有存储（包括加密和非加密）
        DMPStorage.shared.clearAllStorage()
        return true
    }
    
    // Set storage
    @BridgeMethod(SET_STORAGE)
    var setStorage: DMPBridgeMethodHandler = { param, env, callback in
        let param = param.getMap()
        guard let key = param.get("key") as? String else {
            let errMsg = "\(SET_STORAGE):fail missing parameter key"
            DMPContainerApi.invokeFailure(callback: callback, param: nil, errMsg: errMsg)
            return nil
        }
        
        let data = param.get("data")
        let encrypt = param.get("encrypt") as? Bool ?? false
        
        guard let data = data else {
            let errMsg = "\(SET_STORAGE):fail missing parameter data"
            DMPContainerApi.invokeFailure(callback: callback, param: nil, errMsg: errMsg)
            return nil
        }
        
        // 在后台线程执行存储操作
        DispatchQueue.global().async {
            let success = DMPStorage.shared.set(key: key, value: data, encrypted: encrypt)
            
            DispatchQueue.main.async {
                if success {
                    let resultMap = DMPMap()
                    resultMap.set("errMsg", "\(SET_STORAGE):ok")
                    DMPContainerApi.invokeSuccess(callback: callback, param: resultMap)
                } else {
                    DMPContainerApi.invokeFailure(callback: callback, param: nil, errMsg: "\(SET_STORAGE):fail")
                }
            }
        }
        
        return nil
    }
    
    // Get storage
    @BridgeMethod(GET_STORAGE)
    var getStorage: DMPBridgeMethodHandler = { param, env, callback in
        let param = param.getMap()
        guard let key = param.get("key") as? String else {
            let errMsg = "\(GET_STORAGE):fail missing parameter key"
            DMPContainerApi.invokeFailure(callback: callback, param: nil, errMsg: errMsg)
            return nil
        }
        
        let encrypt = param.get("encrypt") as? Bool ?? false
        
        // 在后台线程执行获取操作
        DispatchQueue.global().async {
            let value = DMPStorage.shared.get(key: key, encrypted: encrypt)
            
            DispatchQueue.main.async {
                if let value = value {
                    let resultMap = DMPMap()
                    resultMap.set("data", value)
                    resultMap.set("errMsg", "\(GET_STORAGE):ok")
                    DMPContainerApi.invokeSuccess(callback: callback, param: resultMap)
                } else {
                    DMPContainerApi.invokeFailure(callback: callback, param: nil, errMsg: "\(GET_STORAGE):fail data not found")
                }
            }
        }
        
        return nil
    }
    
    // Remove storage
    @BridgeMethod(REMOVE_STORAGE)
    var removeStorage: DMPBridgeMethodHandler = { param, env, callback in
        let param = param.getMap()
        guard let key = param.get("key") as? String else {
            let errMsg = "\(REMOVE_STORAGE):fail missing parameter key"
            DMPContainerApi.invokeFailure(callback: callback, param: nil, errMsg: errMsg)
            return nil
        }
        
        let encrypt = param.get("encrypt") as? Bool ?? false
        
        // 在后台线程执行删除操作
        DispatchQueue.global().async {
            DMPStorage.shared.remove(key: key, encrypted: encrypt)
            
            DispatchQueue.main.async {
                let resultMap = DMPMap()
                resultMap.set("errMsg", "\(REMOVE_STORAGE):ok")
                DMPContainerApi.invokeSuccess(callback: callback, param: resultMap)
            }
        }
        
        return nil
    }
    
    // Clear storage
    @BridgeMethod(CLEAR_STORAGE)
    var clearStorage: DMPBridgeMethodHandler = { param, env, callback in
        // 在后台线程执行清除操作
        DispatchQueue.global().async {
            DMPStorage.shared.clearAllStorage()
            
            DispatchQueue.main.async {
                let resultMap = DMPMap()
                resultMap.set("errMsg", "\(CLEAR_STORAGE):ok")
                DMPContainerApi.invokeSuccess(callback: callback, param: resultMap)
            }
        }
        
        return nil
    }
    
    // Get storage info synchronously
    @BridgeMethod(GET_STORAGE_INFO_SYNC)
    var getStorageInfoSync: DMPBridgeMethodHandler = { param, env, callback in
        let storageInfo = DMPStorage.shared.getAllStorageInfo()
        
        let result = DMPMap()
        result.set("keys", storageInfo.keys)
        result.set("currentSize", storageInfo.currentSize)
        result.set("limitSize", storageInfo.limitSize)
        
        return result
    }
    
    // Get storage info
    @BridgeMethod(GET_STORAGE_INFO)
    var getStorageInfo: DMPBridgeMethodHandler = { param, env, callback in
        // 在后台线程执行获取存储信息操作
        DispatchQueue.global().async {
            let storageInfo = DMPStorage.shared.getAllStorageInfo()
            
            DispatchQueue.main.async {
                let resultMap = DMPMap()
                resultMap.set("keys", storageInfo.keys)
                resultMap.set("currentSize", storageInfo.currentSize)
                resultMap.set("limitSize", storageInfo.limitSize)
                resultMap.set("errMsg", "\(GET_STORAGE_INFO):ok")
                
                DMPContainerApi.invokeSuccess(callback: callback, param: resultMap)
            }
        }
        
        return nil
    }
}

File: android/dimina/src/main/kotlin/com/didi/dimina/api/storage/StorageApi.kt
package com.didi.dimina.api.storage

import com.didi.dimina.api.APIResult
import com.didi.dimina.api.AsyncResult
import com.didi.dimina.api.BaseApiHandler
import com.didi.dimina.api.NoneResult
import com.didi.dimina.api.SyncResult
import com.didi.dimina.engine.qjs.JSValue
import com.didi.dimina.ui.container.DiminaActivity
import com.tencent.mmkv.MMKV
import org.json.JSONArray
import org.json.JSONObject

/**
 * Storage API implementation
 * Author: Doslin
 *
 * Handles data storage operations like setting, getting, and removing stored data
 */
class StorageApi : BaseApiHandler() {
    companion object {
        private const val SET_STORAGE_SYNC = "setStorageSync"
        private const val GET_STORAGE_SYNC = "getStorageSync"
        private const val REMOVE_STORAGE_SYNC = "removeStorageSync"
        private const val CLEAR_STORAGE_SYNC = "clearStorageSync"
        private const val SET_STORAGE = "setStorage"
        private const val GET_STORAGE = "getStorage"
        private const val REMOVE_STORAGE = "removeStorage"
        private const val CLEAR_STORAGE = "clearStorage"
        private const val GET_STORAGE_INFO_SYNC = "getStorageInfoSync"
        private const val GET_STORAGE_INFO = "getStorageInfo"

        internal fun clearAllStorage(appId: String) {
            MMKV.mmkvWithID(appId).clearAll()
            MMKV.mmkvWithID(StorageRecordCodec.storageId(appId)).clearAll()
        }
    }

    override val apiNames = setOf(
        SET_STORAGE_SYNC,
        GET_STORAGE_SYNC,
        REMOVE_STORAGE_SYNC,
        CLEAR_STORAGE_SYNC,
        SET_STORAGE,
        GET_STORAGE,
        REMOVE_STORAGE,
        CLEAR_STORAGE,
        GET_STORAGE_INFO_SYNC,
        GET_STORAGE_INFO
    )

    override fun handleAction(
        activity: DiminaActivity,
        appId: String,
        apiName: String,
        params: JSONObject,
        responseCallback: (String) -> Unit,
    ): APIResult {
        val legacyStorage = MMKV.mmkvWithID(appId)
        val storage = MMKV.mmkvWithID(StorageRecordCodec.storageId(appId))
        return when (apiName) {
            SET_STORAGE_SYNC -> {
                val kv = params.optJSONArray("args")
                if (kv != null && kv.length() >= 2) {
                    set(kv.getString(0), kv.get(1), storage)
                }
                NoneResult()
            }

            GET_STORAGE_SYNC -> {
                val key = params.optString("args")
                return when (val value = get(key, storage, legacyStorage)) {
                    is String -> SyncResult(JSValue.createString(value))
                    is Int -> SyncResult(JSValue.createNumber(value.toDouble()))
                    is Double -> SyncResult(JSValue.createNumber(value))
                    is Float -> SyncResult(JSValue.createNumber(value.toDouble()))
                    is Long -> SyncResult(JSValue.createNumber(value.toDouble()))
                    is Boolean -> SyncResult(JSValue.createBoolean(value))
                    is JSONArray -> SyncResult(JSValue.createObject(value.toString()))
                    is JSONObject -> SyncResult(JSValue.createObject(value.toString()))
                    else -> SyncResult(JSValue.createString(""))
                }
            }

            REMOVE_STORAGE_SYNC -> {
                val key = params.optString("args")
                if (key.isNotEmpty()) {
                    remove(key, storage)
                }
                NoneResult()
            }

            CLEAR_STORAGE_SYNC -> {
                storage.clear()
                legacyStorage.clear()
                NoneResult()
            }

            SET_STORAGE -> {
                val key = params.optString("key")
                val data = params.opt("data")
                val res = set(key, data, storage)
                AsyncResult(JSONObject().apply {
                    put("errMsg", "$SET_STORAGE:${if (res) "ok" else "fail"}")
                })
            }

            GET_STORAGE -> {
                val key = params.optString("key")
                val value = get(key, storage, legacyStorage)
                AsyncResult(JSONObject().apply {
                    put("data", value)
                    put("errMsg", "$GET_STORAGE:ok")
                })
            }

            REMOVE_STORAGE -> {
                val key = params.optString("key")
                if (key.isNotEmpty()) {
                    remove(key, storage)
                    AsyncResult(JSONObject().apply {
                        put("errMsg", "$REMOVE_STORAGE:ok")
                    })
                } else {
                    AsyncResult(JSONObject().apply {
                        put("errMsg", "$REMOVE_STORAGE:fail")
                    })
                }
            }

            CLEAR_STORAGE -> {
                storage.clear()
                legacyStorage.clear()
                AsyncResult(JSONObject().apply {
                    put("errMsg", "$CLEAR_STORAGE:ok")
                })
            }

            GET_STORAGE_INFO_SYNC -> {
                SyncResult(JSValue.createObject(storageInfo(storage, legacyStorage).toString()))
            }

            GET_STORAGE_INFO -> {
                AsyncResult(storageInfo(storage, legacyStorage).apply {
                    put("errMsg", "$GET_STORAGE_INFO:ok")
                })
            }

            else ->
                super.handleAction(activity, appId, apiName, params, responseCallback)
        }
    }

    private fun set(key: String, data: Any?, storage: MMKV): Boolean {
        if (key.isEmpty()) return false
        return storage.encode(StorageRecordCodec.dataKey(key), StorageRecordCodec.encodeValue(data))
    }

    private fun remove(key: String, storage: MMKV) {
        storage.encode(StorageRecordCodec.dataKey(key), StorageRecordCodec.encodeDeleted())
    }

    private fun get(key: String, storage: MMKV, legacyStorage: MMKV): Any? {
        if (key.isEmpty()) return null
        val dataKey = StorageRecordCodec.dataKey(key)
        if (storage.containsKey(dataKey)) {
            return StorageRecordCodec.decode(storage.decodeString(dataKey))?.takeUnless { it.deleted }?.value
        }
        val legacyValue = getLegacy(key, legacyStorage) ?: return null
        set(key, legacyValue, storage)
        return legacyValue
    }

    private fun getLegacy(key: String, storage: MMKV): Any? {
        if (!storage.containsKey(key)) return null
        val type = storage.decodeString("${key}_type")
            ?: return storage.decodeString(key)
        return when (type) {
            "Int" -> storage.decodeInt(key, 0)
            "String" -> storage.decodeString(key)
            "Boolean" -> storage.decodeBool(key, false)
            "Float" -> storage.decodeFloat(key, 0f)
            "Long" -> storage.decodeLong(key, 0L)
            "Double" -> storage.decodeDouble(key, 0.0)
            "Array" -> storage.decodeString(key)?.let { raw ->
                runCatching { JSONArray(raw) }.getOrElse { raw }
            }
            "Object" -> storage.decodeString(key)?.let { raw ->
                runCatching { JSONObject(raw) }.getOrElse { raw }
            }
            else -> storage.decodeString(key)
        }
    }

    private fun storageInfo(storage: MMKV, legacyStorage: MMKV): JSONObject {
        val values = linkedMapOf<String, Int>()
        val overriddenKeys = mutableSetOf<String>()
        storage.allKeys()?.forEach { storedKey ->
            if (!storedKey.startsWith(StorageRecordCodec.DATA_PREFIX)) return@forEach
            val key = storedKey.removePrefix(StorageRecordCodec.DATA_PREFIX)
            overriddenKeys += key
            val raw = storage.decodeString(storedKey) ?: return@forEach
            val record = StorageRecordCodec.decode(raw) ?: return@forEach
            if (!record.deleted) values[key] = raw.toByteArray().size
        }
        legacyStorage.allKeys()?.forEach { legacyKey ->
            if (legacyKey.endsWith("_type") || legacyKey in overriddenKeys) return@forEach
            val value = getLegacy(legacyKey, legacyStorage) ?: return@forEach
            values[legacyKey] = value.toString().toByteArray().size
        }
        return JSONObject().apply {
            put("keys", JSONArray(values.keys))
            put("currentSize", values.values.sum())
            put("limitSize", 10 * 1024 * 1024)
        }
    }
}

internal data class StorageRecord(val deleted: Boolean, val value: Any? = null)

internal object StorageRecordCodec {
    const val DATA_PREFIX = "data:"
    private const val VERSION = 2

    fun storageId(appId: String): String = "dimina_storage_v2_${appId.length}_$appId"

    fun dataKey(key: String): String = "$DATA_PREFIX$key"

    fun encodeDeleted(): String = JSONObject()
        .put("version", VERSION)
        .put("kind", "deleted")
        .toString()

    fun encodeValue(value: Any?): String {
        val type = when (value) {
            null, JSONObject.NULL -> "Null"
            is Int -> "Int"
            is String -> "String"
            is Boolean -> "Boolean"
            is Float -> "Float"
            is Long -> "Long"
            is Double -> "Double"
            is JSONArray -> "Array"
            is JSONObject -> "Object"
            else -> "String"
        }
        return JSONObject()
            .put("version", VERSION)
            .put("kind", "value")
            .put("type", type)
            .put("value", when (type) {
                "Null" -> JSONObject.NULL
                "String" -> value?.toString() ?: ""
                else -> value
            })
            .toString()
    }

    fun decode(raw: String?): StorageRecord? {
        if (raw == null) return null
        return runCatching {
            val json = JSONObject(raw)
            if (json.optInt("version") != VERSION) return null
            if (json.optString("kind") == "deleted") return StorageRecord(deleted = true)
            if (json.optString("kind") != "value") return null
            val value = when (json.optString("type")) {
                "Null" -> JSONObject.NULL
                "Int" -> json.getInt("value")
                "String" -> json.getString("value")
                "Boolean" -> json.getBoolean("value")
                "Float" -> json.getDouble("value").toFloat()
                "Long" -> json.getLong("value")
                "Double" -> json.getDouble("value")
                "Array" -> json.getJSONArray("value")
                "Object" -> json.getJSONObject("value")
                else -> return null
            }
            StorageRecord(deleted = false, value = value)
        }.getOrNull()
    }
}
